onnxruntime/test/framework/inference_session_test.cc

```diff
@@ -1346,66 +1348,102 @@ TEST(InferenceSessionTests, TestCopyToFromDevices) {
   run_test(run_number++);
 }
 
-TEST(InferenceSessionTests, TestL1Transformers) {
-  string model_uri = "testdata/transform/fusion/fuse-conv-bn-mul-add-unsqueeze.onnx";  
+// This test validates the RegisterTransformer API
+// It creates and registers a dummy transformer and after session initialize
+// validates that this transformer was called regardless of the graph optimization level set.
+TEST(InferenceSessionTests, TestRegisterTransformers) {
+  string model_uri = "testdata/transform/fusion/fuse-conv-bn-mul-add-unsqueeze.onnx";
 
-  SessionOptions so;
-  so.session_logid = "InferenceSessionTests.TestL1Transformers";
-  so.graph_optimization_level = TransformerLevel::Level1;
-  InferenceSession session_object{so, &DefaultLoggingManager()};
-  ASSERT_TRUE(session_object.Load(model_uri).IsOK());
+  for (int i = static_cast<int>(TransformerLevel::Default); i < static_cast<int>(TransformerLevel::MaxTransformerLevel); i++) {
+    SessionOptions so;
+    so.session_logid = "InferenceSessionTests.TestL1AndL2Transformers";
+    so.graph_optimization_level = static_cast<TransformerLevel>(i);
+    InferenceSession session_object{so, &DefaultLoggingManager()};
 
-  std::shared_ptr<Model> p_model;
-  ASSERT_TRUE(Model::Load(model_uri, p_model).IsOK());
+    // Create and register dummy graph transformer
+    auto dummy_transformer_unique_ptr = std::make_unique<DummyGraphTransformer>("DummyTransformer");
+    const auto* dummy_transformer = dummy_transformer_unique_ptr.get();
+    session_object.RegisterGraphTransformer(std::move(dummy_transformer_unique_ptr));
 
-  Status st = session_object.Initialize();
-  ASSERT_TRUE(st.IsOK()) << st;
+    session_object.Load(model_uri);
+    ASSERT_TRUE(session_object.Initialize().IsOK());
+
+    // Validate transformer was called after Session.Initialize
+    ASSERT_TRUE(dummy_transformer->IsTransformerInvoked());
+  }
 }
 
+// This test validates session initialize is successful when all the pre-defined
+// L1 and L2 transformers are enabled.
 TEST(InferenceSessionTests, TestL1AndL2Transformers) {
+  // Models which cover all transformers.
```

Comment:
What is this test really validating? We don't check anything after Initialize, so it's not clear to me what this test would catch that the other ones would miss.

Reply:
This test validates that when all transformers are turned on session initialize is successful. SessionInitialize includes graph transformation, graph resolve... This test catches any issues introduced in the graph by graph transformers... for example earlier transformers which add a new node in the graph were not assigning a provider to this new node... this test was able to catch it because SessionInitialize was failing...